### DISCUSSION TITLE
multi-user taskwarrior usage

### DISCUSSION THREAD

**Discussion by pzybe** [GothenburgBitFactory/taskwarrior#2591] [Ideas]
I was looking through the documentation and wondering if anybody uses taskwarrior/taskserver as a sort of kanban for multiple users.  Perhaps using custom user attributes to add a concept of task ownership or submission, and tying those attributes to PAM, or some other form of user directory.  It would be cool to have a taskserver setup whereby a user could see who owns and is handling what tasks, and that users have a 'lock' on a task or something so that multiple users don't try to own the same task.

**Comment by jaker-dotcom** [comment]
I sure have read that taskd is headed in the direction of supporting multiuser workflows.

> You can create as many organizations as you wish (even one per user), and the purpose is simply to group users together. Future features will utilize this. 

From https://gothenburgbitfactory.github.io/taskserver-setup/

Also in [example workflows](https://taskwarrior.org/docs/workflow.html) someone stated that he uses taskwarrior with a team but I think they sync via git. Check out that site if you want to know more.

**Reply by tbabej** [reply]
This is indeed one of our aspirational long-term goals, however, the current state of the tool does not support any specific multi-user features. That said, we'd welcome suggestions on what workflows / usage pattern you'd like to get supported in the future!

**Reply by jaker-dotcom** [reply]
I'd love to help reach this goal but I just started using the shell, nevermind coding. 

Here are the features I come to think of in a few minutes. 

- unassigned tasks (so just a pool of tasks)
- assign tasks to members (maybe in the form of requests so that one can accept or decline) 
- option to list all tasks from the organization whether assigned or not
- changes (esp. done/delete) to the database should be able to be presented in a timesheet-report style but only with things others have done. Thus it would require to save who made which changes.
- someone with admin rights who can reassign tasks (e.g. when member goes inactive)

**Reply by pzybe** [reply]
`taskd` seems like it already has some of the parameters that would lend itself towards multi-user usage.  Namely that a user with certificates needs to created in order to use the server.  What source of truth users come from though and how they're given permissions I'm guessing would be a point of contention.  a database? PAM? signed committers?  I would think being able to extend the server in the same manner the client could be extended would be the best way to let people decide how to go about it. I don't see any documentation around writing extensions for the server though.

**Comment by fecet** [comment]
Anything happen in 2023?

**Reply by djmitche** [reply]
The work on Taskchampion (https://github.com/GothenburgBitFactory/taskwarrior/discussions/3031) is, among other things, aiming to improve synchronization support.  One of the major reasons not to use `taskd` for anything more than a single person's tasks is that it stores everything unencrypted.  For example, the admin of `freecinc.com` (me) could read the tasks of anyone using that site for synchronization.

However, this thread is about a slightly different question, which is how many people could work on the same set of tasks.  Within the model of Taskwarrior, that would basically entail everyone looking at the same set of tasks, and some way of annotating a task as "assigned" to one person or another.  So, quite possible even currently.  For example, you could tag tasks with `+fecet` and `+ptdel` to indicate assignment.

Note that in any kind of distributed system where progress can be made in multiple replicas simultaneously, there's no way to "lock" something.  Instead, the system must be designed to gracefully handle any conflicts.  For example, in the current Taskwarrior, and in Taskchampion as well, if you mark a task as due on the 3rd from your laptop, and mark the same task as due on the 4th from your desktop, then after a few synchronizations the due date will be the same on both machines[*].  I believe both versions choose the change that was made last.

[*] This isn't quite true -- the current Taskwarrior sync model has some dramatic algorithmic flaws and often gets "out of sync".  But the _intent_ is that it keeps the two in sync!

**Reply by fecet** [reply]
"Assign" task by filter still means everyone sync to the same taskserver is identical and they can visualize tasks "assign" to others, right? That's not a big thing, but if there is a method to allow users to “accept” tasks from server and they can still add their tasks, probably personal by themselves, or one can “accept” tasks from many servers would be great.

**Reply by HammyHavoc** [reply]
With the release of TaskWarrior 3.0 and recommending that users sync via a cloud storage backend, what does that mean for multiple users?

I'm not currently a TaskWarrior user, but I'm strongly considering it.

**Reply by djmitche** [reply]
The cloud-storage sync implementation still performs encryption on the client. Further, the expectation is that each user of Taskwarrior would be syncing to a different "bucket" on the cloud service, so e.g., Alice doesn't have access to Bob's bucket, nor the reverse. So that provides two levels of isolation between users: cloud access permissions, and if those fail client-side encryption prevents Alice and Bob from seeing each other's tasks.

If the question is how to allow multiple users to access the same set of tasks, then they would all use the same credentials and encryption secret. This is very similar to how this would have worked with `taskserver` in 2.x.
